arviz/kde_utils.py

```diff
@@ -0,0 +1,863 @@
+        DCT is desired
+
+    Returns
+    -------
+    output : DTC transformed values
+    """
+
+    x_len = len(x)
+
+    even_increasing = np.arange(0, x_len, 2)
+    odd_decreasing = np.arange(x_len - 1, 0, -2)
+
+    x = np.concatenate((x[even_increasing], x[odd_decreasing]))
+
+    w_1k = np.r_[
+        1,
+        (2 * np.exp(-(0 + 1j) * (np.arange(1, x_len)) * np.pi / (2 * x_len)))
+    ]
+    output = np.real(w_1k * fft(x))
+
+    return output
+
+
+def _fixed_point(t, N, k_sq, a_sq):
+    """
+    Implementation of the function t-zeta*gamma^[l](t) derived from equation (30) in [1]
+
+    References
+    ----------
+    .. [1] Kernel density estimation via diffusion.
+       Z. I. Botev, J. F. Grotowski, and D. P. Kroese.
+       Ann. Statist. 38 (2010), no. 5, 2916--2957.
+    """
+
+    k_sq = np.asfarray(k_sq, dtype=np.float64)
+    a_sq = np.asfarray(a_sq, dtype=np.float64)
+
+    l = 7
+    f = np.sum(np.power(k_sq, l) * a_sq * np.exp(-k_sq * np.pi ** 2 * t))
+    f *= 0.5 * np.pi ** (2.0 * l)
+
+    for j in reversed(range(2, l)):
+        c1 = (1 + 0.5 ** (j + 0.5)) / 3
+        c2 = np.product(np.arange(1.0, 2 * j + 1, 2, dtype=np.float64))
+        c2 /= (np.pi / 2) ** 0.5
+        t_j = np.power((c1 * (c2 / (N * f))), (2.0 / (3.0 + 2.0 * j)))
+        f = np.sum(k_sq ** j * a_sq * np.exp(-k_sq * np.pi ** 2.0 * t_j))
+        f *= 0.5 * np.pi ** (2 * j)
+
+    out = t - (2 * N * np.pi ** 0.5 * f) ** (-0.4)
+    return out
+
+
+def _root(function, N, args, x):
+    # The right bound is at most 0.01
+    found = False
+    N = max(min(1050, N), 50)
+    tol = 10e-12 + 0.01 * (N - 50) / 1000
+
+    while not found:
+        try:
+            bw, res = brentq(function, 0, 0.01, args=args, full_output=True, disp=False)
+            found = True if res.converged else False
+        except ValueError:
+            bw = 0
+            tol *= 2.0
+            found = False
+        if bw <= 0 or tol >= 1:
+            # warnings.warn(
+            #    "Improved Sheather-Jones did not converge as expected. "
+            #    "Using Silverman's rule instead.",
+            #    Warning
+            # )
+            bw = (_bw_silverman(x) / np.ptp(x)) ** 2
+            return bw
+    return bw
+
+
+def _check_type(x):
+    """
+    Checks the input is of the correct type.
+    It only accepts numeric lists/numpy arrays of 1 dimension or something that
+    can be flattened to 1 dimension.
+
+    Parameters
+    ----------
+    x : Object whose type is checked before computing the KDE.
+
+    Returns
+    -------
+    x : 1-D numpy array
+        If no error is thrown, a 1 dimensional array of
+        sample data from the variable for which a density estimate is desired.
+
+    """
+    if not isinstance(x, (list, np.ndarray)):
+        raise ValueError((
+            f"Can't produce a density estimator for {type(x)}.\n"
+            f"Please input a list with numbers or numpy array."
+        ))
```

Comment:
I would remove this. I think it limits the usabiity without too much to gain. What if the input is a dask array? or a pytorch tensor? Moreover, I agree that a tuple is not natural in this context, but arguably, lists are not too natural either. 

Eventually we only want to make sure `x` is a numpy array, having informative error messages is good, but our users already know numpy so I think we can trade a little on this front for flexibility and maintainability.

Reply:
So the idea would be not to check the type of the input and let numpy signal the error, if necessary, when casting to numpy array?

Reply:
yes, numpy is the base library for arrays in python and all the other play well with it, `np.array(ary)` will generally work for dask, jax, xarray, pytorch, tensorflow... objects

Reply:
Great, makes much sense. I'll change it.